### PR TITLE
Add webauthn owner

### PR DESCRIPTION
**Pull request by ChiHaoLu**

### CODE REVIEW

File: app/page/webauthn/devtool.tsx [hunk 1]
```diff
@@ -187,7 +187,20 @@ export function WebAuthnDevTool() {
...
+              credential,
+              (key, value) => {
+                // Filtering out properties
+                if (value instanceof Uint8Array) {
+                  return Buffer.from(value).toString("hex")
+                }
+                return value
+              },
+              2
+            )}
```
Comment (pilagod) [inline app/page/webauthn/devtool.tsx:202]: nit: I saw many places use stringify to log info. It may be worth to extract this as function and apply it to all the places.

File: background/storage.ts [hunk 2]
```diff
@@ -111,6 +125,18 @@ export type PasskeyAccount = {
   address: HexString
 }
 
+export type imAccount = {
+  type: AccountType.imAccount
+  chainId: number
+  address: HexString
+  ecdsaValidator: HexString
+  ownerPrivateKey: HexString
+  webAuthnValidator: HexString
```
Comment (pilagod) [inline background/storage.ts:134]: future: Just a note that we may need a nested structure to model the validator in the future, like the account, for instance:

```ts
type imAccount = {
    // ...
    ownerValidator: {
        type: imAccountValidatorType.ECDSA,
        ownerPrivateKey: 0x1234,
        // ... other fields
    }
}
```

File: packages/account/imAccount/validator/WebAuthnOwner.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    const authnticatorDataJson = {
+      rpIdHash: Buffer.from(authData.rpIdHash).toString("hex"),
+      flags: authData.flags,
+      counter: authData.counter
+    }
+
+    const signature = ethers.AbiCoder.defaultAbiCoder().encode(
+      ["uint256", "uint256"],
+      [webAuthnAuthentication.signature.r, webAuthnAuthentication.signature.s]
+    )
```
Comment (pilagod) [inline packages/account/imAccount/validator/WebAuthnOwner.ts:81]: Q: Shall we keep separated `r` and `s` and let the account decide their encoding way?

**Comment by oneleo** [comment]
Unsure if it's my issue, but unable to connect or transfer with waallet-contract.

File: app/page/webauthn/devtool.tsx (hunk 1 above)
Comment (oneleo) [reply app/page/webauthn/devtool.tsx:202]: I think this could be a separate PR.
Would it be appropriate to add a function named stringify2() in #80 PR?

Comment (pilagod) [reply app/page/webauthn/devtool.tsx:202]: The core point is about the conversion of Uint8Array to hex, but not the 2 spaces indent. If you want to provide a single standard in formatting json, I would suggest to name it as `format`, and also include the Uint8Array as intrinsic replacer.

Comment (oneleo) [reply app/page/webauthn/devtool.tsx:202]: Done! Updated PR!

**Comment by ChiHaoLu** [comment]
> Unsure if it's my issue, but unable to connect or transfer with waallet-contract. How's @ChiHaoLu's .env file set up?

```
PLASMO_PUBLIC_SHIP_NAME=ncc-1701
PLASMO_PUBLIC_SHIELD_FREQUENCY=147
PLASMO_PUBLIC_SITE_URL=https://www.plasmo.com
PLASMO_PUBLIC_EXTENSION_ID=imaadogakjipofgdnmcbeijdmcgedcbp

# PLASMO_PUBLIC_ACCOUNT=0x661b4a3909b486a3da520403ecc78f7a7b683c63
PLASMO_PUBLIC_ACCOUNT_OWNER_PRIVATE_KEY=
PLASMO_PUBLIC_NODE_RPC_URL=http://localhost:8545
PLASMO_PUBLIC_BUNDLER_RPC_URL=http://localhost:3000
PLASMO_PUBLIC_CHAIN_ID=1337
# PLASMO_PUBLIC_PASSKEY_ACCOUNT=0x8506416d9f4a802aeb046e896a5a7ea3895f3f92
PLASMO_PUBLIC_IMACCOUNT=0xc43e7efd75d0fc58eb62c6d574f0943fb53cbceb
PLASMO_PUBLIC_ECDSA_VALIDATOR=0xB7f8BC63BbcaD18155201308C8f3540b07f84F5e
PLASMO_PUBLIC_WEBAUTHN_VALIDATOR=0x9A9f2CCfdE556A7E9Ff0848998Aa4a0CFD8863AE
PLASMO_PUBLIC_IMACCOUNT_CREDENTIAL_ID=yNlhKmmODaytLI5OjmUeEgTBzJU
PLASMO_PUBLIC_IMACCOUNT_PASSKEY_X=38767317671848369117005366613252935360278863387712137689801835511363502284053
PLASMO_PUBLIC_IMACCOUNT_PASSKEY_Y=28664284627359735076229063167185851674000669064097709965842737495497904998863

PLASMO_PUBLIC_VERIFYING_PAYMASTER=0x2279B7A0a67DB372996a5FaB50D91eAA73d2eBe6
PLASMO_PUBLIC_VERIFYING_PAYMASTER_OWNER_PRIVATE_KEY=

CODE=PLASMO-GO-030 # Undefined in extension because it does not start with PLASMO_PUBLIC_
CRX_KEY=MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA0ye66bBwgEeRgJOlE7PtO7jTK9R0ySDlXVUPWPkFt7hDt9hTiu2itexieCW/Sbtc4NCpKJIRVHQUZgr7C2zfZHmC6uLVeb2rba+qzz+5PhwIOP/t7GSpWEJR3MXfrMugQy0pXGBvV3SWFyIdrOhC7HULZT9JVfDcCzS9DGIvZ220VhrBFy41XUuul0AG0RKzuVaeGp/BgD3eIGxedZLFNdth0MKoxqOv8XxvnJ1FPL0jkcmPubJVatxwzI+XdingSQ+BDIOumbBSzXVh/InZbiVNEU0liHCDw/gHsK9EV9frG+hBoVB4V0pXKxbt7aOWXqXzv2WqyDu0oTy+FRvxpQIDAQAB
```

File: packages/account/imAccount/validator/WebAuthnOwner.ts (hunk 3 above)
Comment (ChiHaoLu) [reply packages/account/imAccount/validator/WebAuthnOwner.ts:81]: It's sounds rational.

From the point of Passkey Account, the valid format of signature is (whether in P256Owner or WebAuthnOwner):
```solidity
    const signature = ethers.AbiCoder.defaultAbiCoder().encode(
      [
        "bool",
        "bytes",
        "bool",
        "string",
        "uint256",
        "uint256",
        "uint256",
        "uint256"
      ],
      [
        false,
        Uint8Array.from(Buffer.from(authenticatorData, "hex")),
        true,
        clientDataJson,
        clientDataJsonChallengeIndex,
        clientDataJsonTypeIndex,
        r,
        s
      ]
    )
```

But in imAccount, the valid format of signature is (whether in which validator):
```
ethers.AbiCoder.defaultAbiCoder().encode(
      [
        "address",
        "bytes",
        "(bytes authenticatorFlagsAndSignCount,string postChallengeData)"
      ],
      [validatorAddress, signature, webAuthnInput]
    )
```

The position of **WebAuthnValidator** is like the **PasskeyAccount**:

- The **WebAuthnValidator** and **PasskeyAccount** should construct the valid format of signature. 
- The responsibility of **Owner**(whether in WebAuthnOwner or P256Owner) is return the valid signature, which will be `r` and `s`.

So, the **WebAuthnValidator** and **PasskeyAccount** get the valid params (`r`, `s`, and others requirements) from **Owner**, and construct the valid signature for UserOp.

In short, I think the **Owner** returns `r` and `s` will be more flexible for **WebAuthnValidator** and **PasskeyAccount**.

File: packages/account/imAccount/validator/P256Owner.ts [hunk 4]
```diff
@@ -76,7 +72,8 @@ export class P256Owner implements WebAuthnValidatorOwner {
       counter: Number(this.defaultSignCount)
     }
     return {
-      signature: signature as HexString,
+      r: r,
+      s: s,
```
Comment (pilagod) [inline packages/account/imAccount/validator/P256Owner.ts:76]: nit: Maybe wrap them in one signature field (?

```suggestion
      signature: {
         r,
         s,
      }
```

And if the variable name is same as object field name, in JS/TS it can abbreviate the assignment to only the object field name.

**Review by pilagod — APPROVED** [review]

**Comment by oneleo** [comment]
This PR allows passkey certificate verification and transfers with waallet-dapp on localhost.





【Note :notebook_with_decorative_cover:】Selecting "verifying paymaster" in waallet-dapp for transfers triggers an "provider.ts:176 Uncaught (in promise) Error: Send user operation fail" error, which should be resolved by merging the main branch in the next PR.
